In [1]:
data <- read.csv("https://raw.githubusercontent.com/guru99-edu/R-Programming/master/lahman-batting.csv")

## nth observation
The fonction nth() is complementary to first() and last(). You can access the nth observation within a group with the index to return.

For instance, you can filter only the second year that a team played.

In [4]:
library(dplyr)
data %>%
group_by(teamID) %>%
summarise(second_game = nth(yearID, 2)) %>%
arrange(second_game)

`summarise()` ungrouping output (override with `.groups` argument)


teamID second_game
1   BS1    1871       
2   CH1    1871       
3   CL1    1871       
4   FW1    1871       
5   NY2    1871       
6   PH1    1871       
7   RC1    1871       
8   TRO    1871       
9   WS3    1871       
10  BL1    1872       
11  BR1    1872       
12  BR2    1872       
13  MID    1872       
14  WS4    1872       
15  BL4    1873       
16  ELI    1873       
17  PH2    1873       
18  WS5    1873       
19  CH2    1874       
20  HR1    1874       
21  KEO    1875       
22  NH1    1875       
23  PH3    1875       
24  SL1    1875       
25  SL2    1875       
26  WS6    1875       
27  BSN    1876       
28  CHN    1876       
29  CN1    1876       
30  HAR    1876       
... ...    ...        
120 SLF    1914       
121 NEW    1915       
122 ML1    1953       
123 BAL    1954       
124 KC1    1955       
125 LAN    1958       
126 SFN    1958       
127 LAA    1961       
128 MIN    1961       
129 WS2    1961       
130 HOU    1962       
131 NYN    1962       
132 CAL    1965       
133 ATL    1966       
134 OAK    1968       
135 KCA    1969       
136 MON    1969       
137 SDN    1969       
138 SE1    1969       
139 ML4    1970       
140 TEX    1972       
141 SEA    1977       
142 TOR    1977       
143 COL    1993       
144 FLO    1993       
145 ANA    1997       
146 ARI    1998       
147 MIL    1998       
148 TBA    1998       
149 MIA    2012

### Distinct number of observation
The function n() returns the number of observations in a current group. A closed function to n() is n_distinct(), which count the number of unique values.

In the next example, you add up the total of players a team recruited during the all periods.

In [5]:
data %>%
group_by(teamID) %>%
summarise(number_player = n_distinct(playerID)) %>%
arrange(desc(number_player))

`summarise()` ungrouping output (override with `.groups` argument)


teamID number_player
1   CHN    2051         
2   PHI    2036         
3   SLN    1971         
4   CIN    1912         
5   PIT    1877         
6   CLE    1855         
7   BOS    1783         
8   CHA    1744         
9   DET    1658         
10  NYA    1658         
11  BAL    1046         
12  NYN    1043         
13  BSN     978         
14  SDN     977         
15  OAK     924         
16  TEX     922         
17  LAN     919         
18  NY1     894         
19  SFN     889         
20  HOU     885         
21  ATL     876         
22  BRO     875         
23  WS1     870         
24  KCA     857         
25  MIN     809         
26  SEA     799         
27  PHA     783         
28  SLA     764         
29  TOR     745         
30  MON     620         
... ...    ...          
120 WIL    20           
121 BL4    19           
122 CL1    19           
123 RIC    19           
124 WS6    19           
125 ALT    18           
126 BSP    18           
127 FW1    18           
128 SR1    18           
129 BRP    17           
130 KCN    17           
131 NYP    17           
132 PTP    17           
133 CHP    16           
134 ELI    16           
135 ML3    16           
136 PHP    16           
137 ML2    15           
138 WS5    15           
139 WS4    14           
140 IN1    13           
141 SL2    13           
142 MID    12           
143 MLU    12           
144 PH3    12           
145 SL1    12           
146 CH1    11           
147 KEO    11           
148 RC1    11           
149 SPU    11

### Multiple groups
A summary statistic can be realized among multiple groups.

In [6]:
data %>%
group_by(yearID, teamID) %>%
summarise(mean_games = mean(G)) %>%
arrange(desc(teamID, yearID))

`summarise()` regrouping output by 'yearID' (override with `.groups` argument)


yearID teamID mean_games
1    1884   WSU    19.784314 
2    1891   WS9    33.710526 
3    1886   WS8    29.918919 
4    1887   WS8    54.571429 
5    1888   WS8    47.269231 
6    1889   WS8    40.206897 
7    1884   WS7    22.880000 
8    1875   WS6    13.368421 
9    1873   WS5    23.400000 
10   1872   WS4     7.071429 
11   1871   WS3    19.333333 
12   1872   WS3     6.307692 
13   1961   WS2    47.809524 
14   1962   WS2    59.444444 
15   1963   WS2    48.697674 
16   1964   WS2    58.108108 
17   1965   WS2    56.525000 
18   1966   WS2    51.119048 
19   1967   WS2    60.555556 
20   1968   WS2    55.000000 
21   1969   WS2    61.972222 
22   1970   WS2    55.825000 
23   1971   WS2    55.179487 
24   1901   WS1    67.736842 
25   1902   WS1    63.950000 
26   1903   WS1    59.590909 
27   1904   WS1    54.185185 
28   1905   WS1    54.851852 
29   1906   WS1    52.571429 
30   1907   WS1    43.833333 
...  ...    ...    ...       
2836 2017   ATL    48.42857  
2837 1998   ARI    47.34783  
2838 1999   ARI    50.51163  
2839 2000   ARI    52.97619  
2840 2001   ARI    50.44681  
2841 2002   ARI    55.52381  
2842 2003   ARI    53.95349  
2843 2004   ARI    44.13462  
2844 2005   ARI    55.11628  
2845 2006   ARI    50.71111  
2846 2007   ARI    49.04255  
2847 2008   ARI    55.60976  
2848 2009   ARI    51.13333  
2849 2010   ARI    47.31250  
2850 2011   ARI    43.70588  
2851 2012   ARI    45.16327  
2852 2013   ARI    53.38636  
2853 2014   ARI    43.73077  
2854 2015   ARI    47.86000  
2855 2016   ARI    47.98000  
2856 2017   ARI    52.44444  
2857 1997   ANA    51.65116  
2858 1998   ANA    51.26667  
2859 1999   ANA    49.51111  
2860 2000   ANA    50.48889  
2861 2001   ANA    58.36842  
2862 2002   ANA    57.67500  
2863 2003   ANA    53.16279  
2864 2004   ANA    58.34211  
2865 1884   ALT    12.55556

Code Explanation

* group_by(yearID, teamID): Group by year and team
* summarise(mean_games = mean(G)): Summarize the number of game player
* arrange(desc(teamID, yearID)): Sort the data by team and year

### Filter
Before you intend to do an operation, you can filter the dataset. The dataset starts in 1871, and the analysis does not need the years prior to 1980.

In [7]:
data %>%
filter(yearID > 1980) %>%
group_by(yearID) %>%
summarise(mean_game_year = mean(G))

`summarise()` ungrouping output (override with `.groups` argument)


yearID mean_game_year
1  1981   39.97564      
2  1982   56.55444      
3  1983   56.05268      
4  1984   58.07114      
5  1985   57.17735      
6  1986   56.09440      
7  1987   54.67748      
8  1988   54.16329      
9  1989   53.25349      
10 1990   52.19910      
11 1991   54.37753      
12 1992   54.72983      
13 1993   53.24831      
14 1994   42.78835      
15 1995   45.50998      
16 1996   51.11891      
17 1997   51.86084      
18 1998   51.56278      
19 1999   52.81832      
20 2000   49.72182      
21 2001   51.16580      
22 2002   51.81577      
23 2003   50.87899      
24 2004   51.17533      
25 2005   51.45489      
26 2006   50.35004      
27 2007   50.64477      
28 2008   50.22744      
29 2009   49.73919      
30 2010   50.82670      
31 2011   49.48092      
32 2012   49.37429      
33 2013   49.16111      
34 2014   48.47666      
35 2015   47.46568      
36 2016   47.50573      
37 2017   47.35007

Code Explanation

* filter(yearID > 1980): Filter the data to show only the relevant years (i.e. after 1980)
* group_by(yearID): Group by year
* summarise(mean_game_year = mean(G)): Summarize the data

# Ungroup
Last but not least, you need to remove the grouping before you want to change the level of the computation.

In [8]:
data %>%
filter(HR > 0) %>%
group_by(playerID) %>%
summarise(average_HR_game = sum(HR) / sum(G)) %>%
ungroup() %>%
summarise(total_average_homerun = mean(average_HR_game))

`summarise()` ungrouping output (override with `.groups` argument)


total_average_homerun
1 0.06244721

Code Explanation

* filter(HR >0) : Exclude zero homerun
* group_by(playerID): group by player
* summarise(average_HR_game = sum(HR)/sum(G)): Compute average homerun by player
* ungroup(): remove the grouping
* summarise(total_average_homerun = mean(average_HR_game)): Summarize the data

# Aggregating Data
It is relatively easy to collapse data in R using one or more BY variables and a defined function.

In [9]:
attach(mtcars)

The attach() function in R can be used to make objects within dataframes accessible in R with fewer keystrokes. As an example:

In [12]:
head(mtcars)

mpg  cyl disp hp  drat wt    qsec  vs am gear carb
Mazda RX4         21.0 6   160  110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag     21.0 6   160  110 3.90 2.875 17.02 0  1  4    4   
Datsun 710        22.8 4   108   93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive    21.4 6   258  110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout 18.7 8   360  175 3.15 3.440 17.02 0  0  3    2   
Valiant           18.1 6   225  105 2.76 3.460 20.22 1  0  3    1

In [13]:
aggdata <- aggregate(mtcars,by=list(cyl,vs),
                    FUN = mean,na.rm=TRUE)
print(aggdata)

  Group.1 Group.2      mpg cyl   disp       hp     drat       wt     qsec vs
1       4       0 26.00000   4 120.30  91.0000 4.430000 2.140000 16.70000  0
2       6       0 20.56667   6 155.00 131.6667 3.806667 2.755000 16.32667  0
3       8       0 15.10000   8 353.10 209.2143 3.229286 3.999214 16.77214  0
4       4       1 26.73000   4 103.62  81.8000 4.035000 2.300300 19.38100  1
5       6       1 19.12500   6 204.55 115.2500 3.420000 3.388750 19.21500  1
         am     gear     carb
1 1.0000000 5.000000 2.000000
2 1.0000000 4.333333 4.666667
3 0.1428571 3.285714 3.500000
4 0.7000000 4.000000 1.500000
5 0.0000000 3.500000 2.500000


In [14]:
head(aggdata)

Group.1 Group.2 mpg      cyl disp   hp       drat     wt       qsec     vs
1 4       0       26.00000 4   120.30  91.0000 4.430000 2.140000 16.70000 0 
2 6       0       20.56667 6   155.00 131.6667 3.806667 2.755000 16.32667 0 
3 8       0       15.10000 8   353.10 209.2143 3.229286 3.999214 16.77214 0 
4 4       1       26.73000 4   103.62  81.8000 4.035000 2.300300 19.38100 1 
5 6       1       19.12500 6   204.55 115.2500 3.420000 3.388750 19.21500 1 
  am        gear     carb    
1 1.0000000 5.000000 2.000000
2 1.0000000 4.333333 4.666667
3 0.1428571 3.285714 3.500000
4 0.7000000 4.000000 1.500000
5 0.0000000 3.500000 2.500000

In [17]:
agdata <- aggregate(mtcars,by=list(drat,wt),
                    FUN = mean,na.rm=TRUE)


In [18]:
head(agdata)

Group.1 Group.2 mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
1 3.77    1.513   30.4 4    95.1 113 3.77 1.513 16.90 1  1  5    2   
2 4.93    1.615   30.4 4    75.7  52 4.93 1.615 18.52 1  1  4    2   
3 4.22    1.835   33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1   
4 4.08    1.935   27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1   
5 4.43    2.140   26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2   
6 4.08    2.200   32.4 4    78.7  66 4.08 2.200 19.47 1  1  4    1

In [19]:
attach(CO2)

In [20]:
head(CO2)

Plant Type   Treatment  conc uptake
1 Qn1   Quebec nonchilled  95  16.0  
2 Qn1   Quebec nonchilled 175  30.4  
3 Qn1   Quebec nonchilled 250  34.8  
4 Qn1   Quebec nonchilled 350  37.2  
5 Qn1   Quebec nonchilled 500  35.3  
6 Qn1   Quebec nonchilled 675  39.2

In [22]:
aco <- aggregate(CO2,by =list(uptake,conc),
              FUN = mean, na.rm = TRUE)

Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], 

"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: ret

"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: returning NA"Warning message in mean.default(X[[i]], ...):
"argument is not numeric or logical: ret

In [23]:
head(aco)

Group.1 Group.2 Plant Type Treatment conc uptake
1  7.7    95      NA    NA   NA        95    7.7  
2  9.3    95      NA    NA   NA        95    9.3  
3 10.5    95      NA    NA   NA        95   10.5  
4 10.6    95      NA    NA   NA        95   10.6  
5 11.3    95      NA    NA   NA        95   11.3  
6 12.0    95      NA    NA   NA        95   12.0

In [24]:
attach(iris)

In [25]:
head(iris)

Sepal.Length Sepal.Width Petal.Length Petal.Width Species
1 5.1          3.5         1.4          0.2         setosa 
2 4.9          3.0         1.4          0.2         setosa 
3 4.7          3.2         1.3          0.2         setosa 
4 4.6          3.1         1.5          0.2         setosa 
5 5.0          3.6         1.4          0.2         setosa 
6 5.4          3.9         1.7          0.4         setosa

In [26]:
aggmean = aggregate(iris[,1:4],by=list(iris$Species),FUN=mean, na.rm=TRUE)
aggmean

Group.1    Sepal.Length Sepal.Width Petal.Length Petal.Width
1 setosa     5.006        3.428       1.462        0.246      
2 versicolor 5.936        2.770       4.260        1.326      
3 virginica  6.588        2.974       5.552        2.026

In [27]:
agg_sum = aggregate(iris[,1:4],by=list(iris$Species),FUN=sum, na.rm=TRUE)
agg_sum

Group.1    Sepal.Length Sepal.Width Petal.Length Petal.Width
1 setosa     250.3        171.4        73.1         12.3      
2 versicolor 296.8        138.5       213.0         66.3      
3 virginica  329.4        148.7       277.6        101.3

In [28]:
agg_count = aggregate(iris[,1:4],by=list(iris$Species),FUN=length)
agg_count

Group.1    Sepal.Length Sepal.Width Petal.Length Petal.Width
1 setosa     50           50          50           50         
2 versicolor 50           50          50           50         
3 virginica  50           50          50           50

In [29]:
agg_max = aggregate(iris[,1:4],by=list(iris$Species),FUN=max, na.rm=TRUE)
agg_max

Group.1    Sepal.Length Sepal.Width Petal.Length Petal.Width
1 setosa     5.8          4.4         1.9          0.6        
2 versicolor 7.0          3.4         5.1          1.8        
3 virginica  7.9          3.8         6.9          2.5

In [30]:
agg_min = aggregate(iris[,1:4],by=list(iris$Species),FUN=min, na.rm=TRUE)
agg_min

Group.1    Sepal.Length Sepal.Width Petal.Length Petal.Width
1 setosa     4.3          2.3         1.0          0.1        
2 versicolor 4.9          2.0         3.0          1.0        
3 virginica  4.9          2.2         4.5          1.4